src/System.Security.Cryptography.Pkcs/src/Internal/Cryptography/Helpers.cs

```diff
@@ -256,5 +348,135 @@ private static T Upgrade<T>(Pkcs9AttributeObject basicAttribute) where T : Pkcs9
             enhancedAttribute.CopyFrom(basicAttribute);
             return enhancedAttribute;
         }
+
+        public static byte[] GetSubjectKeyIdentifier(this X509Certificate2 certificate)
+        {
+            Debug.Assert(certificate != null);
+
+            X509Extension extension = certificate.Extensions[Oids.SubjectKeyIdentifier];
+
+            if (extension != null)
+            {
+                AsnReader reader = new AsnReader(extension.RawData, AsnEncodingRules.DER);
+
+                if (reader.TryGetPrimitiveOctetStringBytes(out ReadOnlyMemory<byte> contents))
+                {
+                    return contents.ToArray();
+                }
+
+                throw new CryptographicException();
+            }
+
+            // The Desktop/Windows version of this method use CertGetCertificateContextProperty
+            // with a property ID of CERT_KEY_IDENTIFIER_PROP_ID.
+            //
+            // MSDN says that when there's no extension, this method takes the SHA-1 of the
+            // SubjectPublicKeyInfo block, and returns that.
+            //
+            // https://msdn.microsoft.com/en-us/library/windows/desktop/aa376079%28v=vs.85%29.aspx
+
+#pragma warning disable CA5350 // SHA-1 is required for compat.
+            using (HashAlgorithm hash = SHA1.Create())
+#pragma warning restore CA5350 // Do not use insecure cryptographic algorithm SHA1.
+            {
+                ReadOnlyMemory<byte> publicKeyInfoBytes = GetSubjectPublicKeyInfo(certificate);
+                return hash.ComputeHash(publicKeyInfoBytes.ToArray());
+            }
+        }
+
+        internal static void DigestWriter(IncrementalHash hasher, AsnWriter writer)
+        {
+#if netcoreapp
+            byte[] rented = null;
+            Span<byte> span = stackalloc byte[64];
+            ArrayPool<byte> pool = ArrayPool<byte>.Shared;
+            int bytesWritten;
+
+            while (!writer.TryEncode(span, out bytesWritten))
+            {
+                if (rented != null)
+                {
+                    pool.Return(rented);
+                }
+
+                rented = pool.Rent(span.Length * 2);
+                span = rented;
+            }
+
+            hasher.AppendData(span.Slice(0, bytesWritten));
+#else
+            // If there's no AppendData(ReadOnlySpan) just call Encode to get the
+            // properly sized array.
+            hasher.AppendData(writer.Encode());
```

Comment:
Was the netcoreapp method added for performance reasons? Seems like the resizes (even using renting) would be costly unless output sizes are generally pretty small enough to require at most a few doublings.

Reply:
I double checked my math on this, and 64 was almost always too small.  Then I rewrote this.  Then I rewrote it again.  Then I rewrote it again.  Then I realized that the right answer was that the writer could just return a ReadOnlySpan from Encode.  That got messy, so I made EncodeAsSpan.  When we go to make it public later we can decide what we really want to do.